File: Data Structures & Algorithms/balanced-binary-tree/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    int height(TreeNode* node) {
        if (node == nullptr) return 0;

        int lh = height(node->left);
        if (lh == -1) return -1;

        int rh = height(node->right);
        if (rh == -1) return -1;

        if (abs(lh - rh) > 1) return -1;

        return 1 + max(lh, rh);
    }
public:
    bool isBalanced(TreeNode* root) {
        if (height(root) == -1){
            return false;
        }
        return true;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if root is None:
            return 0
        
        good_nodes = [0]

        def dfs(node, prev):
            if node is None:
                return
            print(node.val, prev)
            if node.val >= prev:
                good_nodes[0] += 1
            dfs(node.left, max(node.val, prev))
            dfs(node.right, max(node.val, prev))
        
        dfs(root, root.val)
        return good_nodes[0]
        